### ISSUE TITLE
Incorrect loss_cls after turn on fp16

### ISSUE THREAD

**Issue by zlin3000** [open-mmlab/mmdetection#5090]
Thanks for your error report and we appreciate it a lot.

**Checklist**

1. I have searched related issues but cannot get the expected help.
2. I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
3. The bug has not been fixed in the latest version.

**Describe the bug**
while turning on fp16, the value of loss_cls is incorrect, which is about 10000+ depends on the different backbones and detection methods(I test FCOS and GFL with resnet50 and resnest50). 
fp32 works fine.

**Reproduction**

1. What command or script did you run?

```none
A placeholder for the command.
```

2. Did you make any modifications on the code or config? Did you understand what you have modified?

3. What dataset did you use?

**Environment**
mmdetection(20210430) master code
mmcv-full == 1.3.2

**Error traceback**
If applicable, paste the error trackback here.

```none
A placeholder for trackback.
```

**Bug fix**
If you have already identified the reason, you can provide the information here. If you are willing to create a PR to fix it, please also leave a comment here and that would be much appreciated!

**Comment by shinya7y** [comment]
Could you paste the config that shows incorrect loss_cls?

**Comment by ZwwWayne** [comment]
Please provide the config and log for use to reproduce the error. We are not able to help here without more information.

**Comment by zlin3000** [comment]
Sorry for the late, here is the info 
**part of config:**
norm_cfg = dict(type='SyncBN', requires_grad=True)
model = dict(
    type='GFL',
    pretrained='open-mmlab://resnest50',
    backbone=dict(
        type='ResNeSt',
        stem_channels=64,
        depth=50,
        radix=2,
        reduction_factor=4,
        avg_down_stride=True,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='SyncBN', requires_grad=True),
        norm_eval=False,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs='on_output',
        num_outs=5),
    bbox_head=dict(
        type='GFLHead',
        num_classes=20,
        in_channels=256,
        stacked_convs=4,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            ratios=[1.0],
            octave_base_scale=8,
            scales_per_octave=1,
            strides=[8, 16, 32, 64, 128]),
        loss_cls=dict(
            type='QualityFocalLoss',
            use_sigmoid=True,
            beta=2.0,
            loss_weight=1.0),
        loss_dfl=dict(type='DistributionFocalLoss', loss_weight=0.25),
        reg_max=16,
        loss_bbox=dict(type='GIoULoss', loss_weight=2.0)),
    train_cfg=dict(
        assigner=dict(type='ATSSAssigner', topk=9),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    test_cfg=dict(
        nms_pre=1000,
        min_bbox_size=0,
        score_thr=0.05,
        nms=dict(type='nms', iou_threshold=0.6),
        max_per_img=100))
optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)
fp16 = dict(loss_scale=512.0)

**part of log：**
2021-05-06 14:24:02,878 - mmdet - INFO - Epoch [1][50/13445]	lr: 3.463e-04, eta: 11 days, 4:00:45, time: 1.994, data_time: 0.340, memory: 4671, loss_cls: 8263.5787, loss_bbox: 1.4732, loss_dfl: 0.7296, loss: 8265.7815
INFO:mmdet:Epoch [1][50/13445]	lr: 3.463e-04, eta: 11 days, 4:00:45, time: 1.994, data_time: 0.340, memory: 4671, loss_cls: 8263.5787, loss_bbox: 1.4732, loss_dfl: 0.7296, loss: 8265.7815
2021-05-06 14:25:25,230 - mmdet - INFO - Epoch [1][100/13445]	lr: 6.793e-04, eta: 10 days, 4:41:27, time: 1.647, data_time: 0.011, memory: 4671, loss_cls: 8721.7952, loss_bbox: 1.4586, loss_dfl: 0.7353, loss: 8723.9891
INFO:mmdet:Epoch [1][100/13445]	lr: 6.793e-04, eta: 10 days, 4:41:27, time: 1.647, data_time: 0.011, memory: 4671, loss_cls: 8721.7952, loss_bbox: 1.4586, loss_dfl: 0.7353, loss: 8723.9891
2021-05-06 14:26:47,951 - mmdet - INFO - Epoch [1][150/13445]	lr: 1.012e-03, eta: 9 days, 21:13:58, time: 1.654, data_time: 0.010, memory: 4671, loss_cls: 8842.0744, loss_bbox: 1.4548, loss_dfl: 0.7293, loss: 8844.2584
INFO:mmdet:Epoch [1][150/13445]	lr: 1.012e-03, eta: 9 days, 21:13:58, time: 1.654, data_time: 0.010, memory: 4671, loss_cls: 8842.0744, loss_bbox: 1.4548, loss_dfl: 0.7293, loss: 8844.2584
2021-05-06 14:28:09,991 - mmdet - INFO - Epoch [1][200/13445]	lr: 1.345e-03, eta: 9 days, 17:02:04, time: 1.641, data_time: 0.010, memory: 4671, loss_cls: 9308.0295, loss_bbox: 1.5013, loss_dfl: 0.7281, loss: 9310.2589
INFO:mmdet:Epoch [1][200/13445]	lr: 1.345e-03, eta: 9 days, 17:02:04, time: 1.641, data_time: 0.010, memory: 4671, loss_cls: 9308.0295, loss_bbox: 1.5013, loss_dfl: 0.7281, loss: 9310.2589
2021-05-06 14:29:32,966 - mmdet - INFO - Epoch [1][250/13445]	lr: 1.678e-03, eta: 9 days, 15:00:31, time: 1.659, data_time: 0.010, memory: 4671, loss_cls: 9402.4944, loss_bbox: 1.4672, loss_dfl: 0.7329, loss: 9404.6944
INFO:mmdet:Epoch [1][250/13445]	lr: 1.678e-03, eta: 9 days, 15:00:31, time: 1.659, data_time: 0.010, memory: 4671, loss_cls: 9402.4944, loss_bbox: 1.4672, loss_dfl: 0.7329, loss: 9404.6944

**Comment by shinya7y** [comment]
I couldn't reproduce the issue on COCO.

PyTorch: 1.7.1
MMCV: 1.3.2
MMDetection: 2.12.0+931d96e

```python
_base_ = [
    '../_base_/datasets/coco_detection.py',
    '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
]

norm_cfg = dict(type='SyncBN', requires_grad=True)
model = dict(
    type='GFL',
    pretrained='open-mmlab://resnest50',
    backbone=dict(
        type='ResNeSt',
        stem_channels=64,
        depth=50,
        radix=2,
        reduction_factor=4,
        avg_down_stride=True,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=norm_cfg,
        norm_eval=False,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=1,
        add_extra_convs='on_output',
        num_outs=5),
    bbox_head=dict(
        type='GFLHead',
        num_classes=80,
        in_channels=256,
        stacked_convs=4,
        feat_channels=256,
        anchor_generator=dict(
            type='AnchorGenerator',
            ratios=[1.0],
            octave_base_scale=8,
            scales_per_octave=1,
            strides=[8, 16, 32, 64, 128]),
        loss_cls=dict(
            type='QualityFocalLoss',
            use_sigmoid=True,
            beta=2.0,
            loss_weight=1.0),
        loss_dfl=dict(type='DistributionFocalLoss', loss_weight=0.25),
        reg_max=16,
        loss_bbox=dict(type='GIoULoss', loss_weight=2.0)),
    train_cfg=dict(
        assigner=dict(type='ATSSAssigner', topk=9),
        allowed_border=-1,
        pos_weight=-1,
        debug=False),
    test_cfg=dict(
        nms_pre=1000,
        min_bbox_size=0,
        score_thr=0.05,
        nms=dict(type='nms', iou_threshold=0.6),
        max_per_img=100))

optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)

fp16 = dict(loss_scale=512.0)
```

```
2021-05-07 00:24:03,103 - mmdet - INFO - Epoch [1][50/58633]    lr: 1.978e-03, eta: 19 days, 14:43:01, time: 2.409, data_time: 0.885, memory: 3183, loss_cls: 0.6240, loss_bbox: 1.1906, loss_dfl: 0.6241, loss: 2.4387
2021-05-07 00:25:12,367 - mmdet - INFO - Epoch [1][100/58633]   lr: 3.976e-03, eta: 15 days, 10:41:38, time: 1.385, data_time: 0.005, memory: 3183, loss_cls: 0.7153, loss_bbox: 1.1443, loss_dfl: 0.5479, loss: 2.4075
2021-05-07 00:26:22,402 - mmdet - INFO - Epoch [1][150/58633]   lr: 5.974e-03, eta: 14 days, 2:20:43, time: 1.401, data_time: 0.005, memory: 3183, loss_cls: 0.6680, loss_bbox: 1.1246, loss_dfl: 0.5397, loss: 2.3323
2021-05-07 00:27:30,713 - mmdet - INFO - Epoch [1][200/58633]   lr: 7.972e-03, eta: 13 days, 8:28:33, time: 1.366, data_time: 0.005, memory: 3183, loss_cls: 0.6854, loss_bbox: 1.0567, loss_dfl: 0.5228, loss: 2.2649
```

**Comment by zlin3000** [comment]
> I couldn't reproduce the issue on COCO.
> 
> PyTorch: 1.7.1
> MMCV: 1.3.2
> MMDetection: 2.12.0+931d96e
> 
> ```python
> _base_ = [
>     '../_base_/datasets/coco_detection.py',
>     '../_base_/schedules/schedule_1x.py', '../_base_/default_runtime.py'
> ]
> 
> norm_cfg = dict(type='SyncBN', requires_grad=True)
> model = dict(
>     type='GFL',
>     pretrained='open-mmlab://resnest50',
>     backbone=dict(
>         type='ResNeSt',
>         stem_channels=64,
>         depth=50,
>         radix=2,
>         reduction_factor=4,
>         avg_down_stride=True,
>         num_stages=4,
>         out_indices=(0, 1, 2, 3),
>         frozen_stages=1,
>         norm_cfg=norm_cfg,
>         norm_eval=False,
>         style='pytorch'),
>     neck=dict(
>         type='FPN',
>         in_channels=[256, 512, 1024, 2048],
>         out_channels=256,
>         start_level=1,
>         add_extra_convs='on_output',
>         num_outs=5),
>     bbox_head=dict(
>         type='GFLHead',
>         num_classes=80,
>         in_channels=256,
>         stacked_convs=4,
>         feat_channels=256,
>         anchor_generator=dict(
>             type='AnchorGenerator',
>             ratios=[1.0],
>             octave_base_scale=8,
>             scales_per_octave=1,
>             strides=[8, 16, 32, 64, 128]),
>         loss_cls=dict(
>             type='QualityFocalLoss',
>             use_sigmoid=True,
>             beta=2.0,
>             loss_weight=1.0),
>         loss_dfl=dict(type='DistributionFocalLoss', loss_weight=0.25),
>         reg_max=16,
>         loss_bbox=dict(type='GIoULoss', loss_weight=2.0)),
>     train_cfg=dict(
>         assigner=dict(type='ATSSAssigner', topk=9),
>         allowed_border=-1,
>         pos_weight=-1,
>         debug=False),
>     test_cfg=dict(
>         nms_pre=1000,
>         min_bbox_size=0,
>         score_thr=0.05,
>         nms=dict(type='nms', iou_threshold=0.6),
>         max_per_img=100))
> 
> optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)
> 
> fp16 = dict(loss_scale=512.0)
> ```
> 
> ```
> 2021-05-07 00:24:03,103 - mmdet - INFO - Epoch [1][50/58633]    lr: 1.978e-03, eta: 19 days, 14:43:01, time: 2.409, data_time: 0.885, memory: 3183, loss_cls: 0.6240, loss_bbox: 1.1906, loss_dfl: 0.6241, loss: 2.4387
> 2021-05-07 00:25:12,367 - mmdet - INFO - Epoch [1][100/58633]   lr: 3.976e-03, eta: 15 days, 10:41:38, time: 1.385, data_time: 0.005, memory: 3183, loss_cls: 0.7153, loss_bbox: 1.1443, loss_dfl: 0.5479, loss: 2.4075
> 2021-05-07 00:26:22,402 - mmdet - INFO - Epoch [1][150/58633]   lr: 5.974e-03, eta: 14 days, 2:20:43, time: 1.401, data_time: 0.005, memory: 3183, loss_cls: 0.6680, loss_bbox: 1.1246, loss_dfl: 0.5397, loss: 2.3323
> 2021-05-07 00:27:30,713 - mmdet - INFO - Epoch [1][200/58633]   lr: 7.972e-03, eta: 13 days, 8:28:33, time: 1.366, data_time: 0.005, memory: 3183, loss_cls: 0.6854, loss_bbox: 1.0567, loss_dfl: 0.5228, loss: 2.2649
> ```

Could u try this with VOC? I used a custom dataset with  VOC format.

**Comment by shinya7y** [comment]
COCO training in your environment will be easier and more informative.
Training on val2017 is sufficient for debug.
